File: authentication.swift
import Foundation
import CryptoKit 

struct User: Codable {
    let username: String
    let password: String  
}

let usersFile = "users.json"  

func loadUsers() -> [User] {
    guard let url = Bundle.main.url(forResource: "users", withExtension: "json") else {
        print("users.json file not found!")
        return []
    }

    do {
        let data = try Data(contentsOf: url)
        let users = try JSONDecoder().decode([User].self, from: data)
        return users
    } catch {
        print("Error loading JSON: \(error)")
        return []
    }
}

func saveUsers(users: [User]) {
    let fileURL = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0].appendingPathComponent("users.json")

    do {
        let data = try JSONEncoder().encode(users)
        try data.write(to: fileURL)
        print("Users saved successfully.")
    } catch {
        print("Error saving JSON: \(error)")
    }
}

func hashPassword(_ password: String) -> String {
    let inputData = Data(password.utf8)
    let hashed = SHA256.hash(data: inputData)
    return hashed.compactMap { String(format: "%02x", $0) }.joined()
}

func registerUser(username: String, password: String) {
    var users = loadUsers()

    if users.contains(where: { $0.username == username }) {
        print("Username already exists")
        return
    }

    let newUser = User(username: username, password: hashPassword(password))
    users.append(newUser)

    saveUsers(users: users)
    print("Registration successful!")
}

func authenticateUser(username: String, password: String) -> Bool {
    let users = loadUsers()
    let hashedPassword = hashPassword(password)

    return users.contains { $0.username == username && $0.password == hashedPassword }
}

print("Choose an option:")
print("1️. Register")
print("2️. Sign In")

if let choice = readLine() {
    if choice == "1" {
        print("Choose a username:")
        let username = readLine() ?? ""
        print("Choose a password:")
        let password = readLine() ?? ""

        registerUser(username: username, password: password)
    } else if choice == "2" {
        print("Enter username:")
        let username = readLine() ?? ""
        print("Enter password:")
        let password = readLine() ?? ""

        if authenticateUser(username: username, password: password) {
            print("Login Successful!")
        } else {
            print("Invalid Credentials.")
        }
    } 
}

File: backend/authentication.py
import json
import hashlib
import os

USER_FILE = "users.json"

def load_users():
    if not os.path.exists(USER_FILE):  
        with open(USER_FILE, "w") as file:
            json.dump({"users": []}, file)
    with open(USER_FILE, "r") as file:
        data = json.load(file)
    return data["users"]

def save_users(users):
    with open(USER_FILE, "w") as file:
        json.dump({"users": users}, file, indent=4)

def hash_password(password):
    return hashlib.sha256(password.encode()).hexdigest()

def register_user(username, password):
    users = load_users()

    if any(user["username"] == username for user in users):
        print("Username already exists")
        return

    users.append({"username": username, "password": hash_password(password)})
    save_users(users)
    print("Registration successful")

def authenticate_user(username, password):
    users = load_users()
    hashed_input_password = hash_password(password)  

    for user in users:
        if user["username"] == username and user["password"] == hashed_input_password:
            return True  
    return False 

def main():
    print("Welcome! Choose an option:")
    print("1️. Register")
    print("2️. Sign In")
    
    choice = input("Enter 1 or 2: ")

    if choice == "1":
        username = input("Choose a username: ")
        password = input("Choose a password: ")
        register_user(username, password)
    elif choice == "2":
        username = input("Enter username: ")
        password = input("Enter password: ")

        if authenticate_user(username, password):
            print("Login Successful!")
        else:
            print(" Invalid Credentials.")

if __name__ == "__main__":
    main()
